src/cart/crt_iv.c

```diff
@@ -3291,6 +3290,9 @@ crt_iv_update_internal(crt_iv_namespace_t ivns, uint32_t class_id,
 		if (sync_type.ivs_flags & CRT_IV_SYNC_BIDIRECTIONAL) {
 			rc = update_comp_cb(ivns_internal, class_id, iv_key,
 					    NULL, iv_value, rc, cb_arg);
+			if (rc == 0 && sync_type.ivs_comp_cb)
+				sync_type.ivs_comp_cb(sync_type.ivs_comp_cb_arg,
+						      rc);
```

Comment:
what is the reasoning for this? the documentation already states that sync is not to be set when BIDIRECTIONAL flag is specified:
"594          * Currently, if this flag is specified ivs_mode must be set to
595          * CRT_IV_SYNC_NONE, ivs_event must be set to CRT_IV_SYNC_EVENT_UPDATE
596          */
"

Reply:
I just traced resources which were allocated in engine/server_iv.c in L1027-L1051 and passed to here. I observed they will be lost here. If you think this is wrong how to clean those resources? We have to put "iv_ns", free SGL iv_value and free cb_arg.

Reply:
the update_comp_cb is the one that should be doing whatever cleanup it needs as its an indication that update has finished and it is the last stage of operation that IV framework performs in this code path on this node.

in this case when there is no synchronization calling sync completion callback is not something that should be done.
It sounds like daos does register a sync with this, but i am not sure what is the intention here.

@wangdi1 can you comment on this?

Reply:
daos will only register the iv_sync comp callback for lazy sync mode to release the buffer asynchronously.  see iv_op()
if (sync && sync->ivs_mode == CRT_IV_SYNC_LAZY) {

So if this BIDIRECTIONAL mode have to use with SYNC_NONE, then maybe it should be an assertion here?

D_ASSERT(sync_type.ivs_comp_cb == NULL);

Reply:
done